cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -363,12 +364,64 @@ public ApplicationWithPrograms updateApp(ApplicationId appId, AppRequest appRequ
                      programTerminator, ownerAdmin.getOwner(appId), appRequest.canUpdateSchedules());
   }
 
+  /**
+   * Finds latest application artifact for given application and current artifact for upgrading application.
+   *
+   * @param appId application Id to find latest app artifact for.
+   * @param currentArtifactId current artifact used by application.
+   * @param artifactScopesToConsider artifact scopes to search in for finding candidate artifacts.
+   * @param considerSnapshot whether to consider snapshot version of artifacts or not for upgrade.
+   * @return {@link ArtifactSummary} for the artifact to be used for upgrade purpose.
+   * @throws NotFoundException if there is no artifact found for upgrade. This should not happen in normal case.
+   * @throws Exception if there was an exception during finding candidate artifact.
+   */
+  private ArtifactSummary getLatestAppArtifactForUpgrade(ApplicationId appId, ArtifactId currentArtifactId,
+                                                         Set<ArtifactScope> artifactScopesToConsider,
+                                                         boolean considerSnapshot)
+    throws Exception {
+
+    List<ArtifactSummary> availableArtifacts = new ArrayList<>();
+    // Find candidate artifacts from all scopes we need to consider.
+    for (ArtifactScope scope: artifactScopesToConsider) {
+      NamespaceId artifactNamespaceToConsider =
+        ArtifactScope.SYSTEM.equals(scope) ? NamespaceId.SYSTEM : appId.getParent();
+      List<ArtifactSummary> artifacts = null;
+      try {
+        artifacts = artifactRepository.getArtifactSummaries(artifactNamespaceToConsider, currentArtifactId.getName(),
+                                                            Integer.MAX_VALUE, ArtifactSortOrder.ASC);
+      } catch (ArtifactNotFoundException e) {
+        // This can happen if we are looking for candidate artifact in multiple namespace.
+        continue;
+      }
+      for (ArtifactSummary artifactSummary: artifacts) {
+        ArtifactVersion artifactVersion = new ArtifactVersion(artifactSummary.getVersion());
+        // Consider if it is a non-snapshot version artifact or it is a snapshot version than considerSnapshot is true.
+        if ((artifactVersion.isSnapshot() && considerSnapshot) || !artifactVersion.isSnapshot()) {
+          availableArtifacts.add(artifactSummary);
+        }
+      }
+    }
+    // Find the artifact with latest version.
+    Optional<ArtifactSummary> newArtifactCandidate = availableArtifacts.stream().max(
+      Comparator.comparing(artifactSummary -> new ArtifactVersion(artifactSummary.getVersion())));
+
+    if (!newArtifactCandidate.isPresent()) {
+      String error = String.format("No artifacts found for artifact %s for app %s.",
+                                   currentArtifactId.getName(), appId);
+      throw new NotFoundException(error);
```

Comment:
Throw with `ArtifactNotFoundException` exception with the `ArtifactId` in the constructor. The message is already formatted by the exception, no need to use a custom message.